src/structs.jl

```diff
@@ -428,7 +428,33 @@ end
     return x_i
 end
 
-@inline function read(::Struct, buf, pos, len, b, ::Type{T}; kw...) where {T}
+mutable struct UnorderedStructClosure{T, KW}
+    buf::T
+    pos::Int
+    len::Int
+    kw::KW
+end
+@inline function (f::UnorderedStructClosure)(T)
+    pos, key, value = readkeyvalue(f.buf, f.pos, f.len, T; f.kw...)
+    f.pos = pos
+    return key, value
+end
+
+## TODO: delete on of thsee and move some vesrion to StructTypes?
+@generated function construct_nt(::Type{T}, vals::NamedTuple) where {T}    
+    args = map(fieldnames(T)) do name                                   
+         :(vals.$name)                                                  
+    end                                                                 
+    return Expr(:block,                                                 
+                Expr(:meta, :inline),                                   
+                Expr(:call, T, args...))                                
+end
+function construct_nv(::Type{T}, nv::Tuple) where {T}
+    reordered_vals = ntuple(i -> nv[Base.fieldindex(T, nv[i][1])][2], fieldcount(T))
```

Comment:
This isn't quite right; for example, if you have a struct like:
```julia
struct T
    a::Int
    b::Float64
    c::String
end
```
But read json like `{"c": "hey", "a": 1, "b:" 3.14}`, then the fields get mixed up.